### PR TITLE
Basic Go Programs: Day 10-12

### PR DESCRIPTION
**Pull request by mrniranjan**
Loops: 2 D Arrays

Signed-off-by: Niranjan MR <niranjan@ashoo.in>

### CODE REVIEW

File: basics/list_subdirectory.go [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+	"path/filepath"
+)
+
+func main() {
+	dir := "/tmp"
+	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		fmt.Println("files :", path)
+		return nil
+	})
+	if err != nil {
```
Comment (gsr-shanks) [inline basics/list_subdirectory.go:21]: Check the example here https://play.golang.org/p/Uk5VngvDCb1
err should be returned immediately before anything

File: loops/loop49.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+func main() {
+	m1 := [2][2]int{ {1, 2}, {3, 4}	}
+	m2 := [2][2]int{ {3, 1}, {4, 2} }
+	m3 := [2][2]int{}
+	fmt.Println(m1)
+	fmt.Println(m2)
+	for i:=0; i < 2; i++ {
+		for j:=0; j < 2; j++ {
+			m3[i][j] = m1[i][j] + m2[i][j]
+			fmt.Printf("%d ", m3[i][j])
```
Comment (gsr-shanks) [inline loops/loop49.go:17]: Received: 
4 3 
7 6 

Expected:
4 6 
7 3

File: loops/loop50.go [hunk 3]
```diff
@@ -0,0 +1,19 @@
+/* Making 2D Slices */
+package main
+
+import (
+	"fmt"
+)
+
+func main() {
+	n := 2 // 2 rows
+	m := 3 // 3 colums
```
Comment (gsr-shanks) [inline loops/loop50.go]: typo: columns

File: loops/loop54.go [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+                the Body stream*/
+		defer resp.Body.Close()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "fetch: reading%s: %v\n", url, err)
+			os.Exit(1)
+		}
+		fmt.Printf("%s", b)
+	}
+		
+
```
Comment (gsr-shanks) [inline loops/loop54.go:23]: blank lines

File: loops/loop54.go [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+		resp, err := http.Get(url)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "fetch : %v\n", err)
+			os.Exit(1)
+		}
+		/*ioutil reads the entire response , the result is stored in b */
+		b, err := ioutil.ReadAll(resp.Body)
+		/* we are using defer to make sure that in case any errors we still close
+                the Body stream*/
+		defer resp.Body.Close()
```
Comment (gsr-shanks) [inline loops/loop54.go]: IDE is complaining: "Possible resource leak, 'defer' is called in a 'for' loop"

add a comment on why there is no leak here.

File: loops/loop56.go [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+func main() {
+	for _, filePaths := range os.Args[1:] {
+		err := filepath.Walk(filePaths, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				fmt.Printf("Something wrong with accessing path %q, %v\n", filePaths, err)
+			}
+			fmt.Println("files: ", path)
+			return nil
+		})
+		if err != nil {
```
Comment (gsr-shanks) [inline loops/loop56.go]: 'err != nil' has been checked already at line 16. Will this check ever happen?

File: functions/fun2.go [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+package main
+
+import (
+	"fmt"
+)
+
+func main() {
+	numbers := []int{328, 4972, 18, 47281, 530}
+	a := bigNum(&numbers)
+	b := bigNum(&numbers)
```
Comment (gsr-shanks) [inline functions/fun2.go]: typo: s/bigNum/curSmall

File: functions/fun4.go [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+func main() {
+	var num int = 53121
+	fmt.Printf("%d has %d digits\n", numberOfDigits(&num), num)
+}
+
+func numberOfDigits(p * int) int {
+	count := 0
+	temp := *p
+	for temp != 0 {
+		digit := temp % 10
```
Comment (gsr-shanks) [inline functions/fun4.go]: digit declared and not used

File: functions/fun8.go [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+	"fmt"
+)
+
+func main() {
+	num := 53323
+	digit(&num)
+}
+
+func digit(num *int) {
+	temp_num := *num
```
Comment (gsr-shanks) [inline functions/fun8.go]: Use camelcase instead of snake case

File: functions/fun10.go [hunk 10]
```diff
@@ -0,0 +1,27 @@
+/* using flag package and accessing flag arguments
+using pointers 
+./fun10 a bc def
+./fun10 -s / a bc def ga
+a/bc/def/ga
+./fun10 --help
```
Comment (gsr-shanks) [inline functions/fun10.go:6]: --help returns non-zero exit. why?

File: functions/fun11.go [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+package main
+
+import (
+	"fmt"
+	"os"
+)
+
+func main() {
+	var m map[string]int
+	if m == nil {
```
Comment (gsr-shanks) [inline functions/fun11.go:12]: Condition is always 'true' because 'm' is always 'nil'

File: loops/loop49.go (hunk 2 above)
Comment (mrniranjan) [reply loops/loop49.go:17]: check  https://paste.centos.org/view/061b68b7

**Review by mrniranjan — COMMENTED** [review]
fixed in updated commit

File: functions/fun10.go (hunk 10 above)
Comment (mrniranjan) [reply functions/fun10.go:6]: From the documentation , this seems a default behaviour: https://golang.org/pkg/flag/#pkg-variables 
https://golang.org/pkg/flag/#ExitOnError (constant with value as 2 )

File: loops/loop49.go (hunk 2 above)
Comment (gsr-shanks) [reply loops/loop49.go:17]: In your paste you have 
`1       2               3       1`
`3       4               4       2`

But in the code you have
`{1, 2}, {3, 4}`
`{3, 1}, {4, 2}`

**Pull request merged**